contracts/token/ERC20/extensions/ERC7540.sol

```diff
@@ -0,0 +1,130 @@
+
+        requestId = _generateRequestId(controller, assets);
+
+        IERC20(asset()).safeTransferFrom(owner, address(this), assets);
+
+        _pendingDepositRequests[controller][requestId].amount += assets;
+
+        emit DepositRequest(controller, owner, requestId, msg.sender, assets);
+    }
+
+    /**
+     * @dev Creates a new redeem request.
+     */
+    function requestRedeem(
+        uint256 shares,
+        address controller,
+        address owner
+    ) external override nonReentrant returns (uint256 requestId) {
+        require(shares > 0, "ERC7540: shares must be greater than zero");
+        require(owner == msg.sender || isOperator(owner, msg.sender), "ERC7540: unauthorized");
+
+        requestId = _generateRequestId(controller, shares);
+
+        _burn(owner, shares);
+
+        _pendingRedeemRequests[controller][requestId].amount += shares;
+
+        emit RedeemRequest(controller, owner, requestId, msg.sender, shares);
+    }
+
+    /**
+     * @dev Gets the pending deposit request amount.
+     */
+    function pendingDepositRequest(uint256 requestId, address controller) external view override returns (uint256) {
+        return _pendingDepositRequests[controller][requestId].amount;
+    }
+
+    /**
+     * @dev Gets the claimable deposit request amount.
+     */
+    function claimableDepositRequest(uint256 requestId, address controller) external view override returns (uint256) {
+        return _pendingDepositRequests[controller][requestId].claimable;
+    }
+
+    /**
+     * @dev Gets the pending redeem request amount.
+     */
+    function pendingRedeemRequest(uint256 requestId, address controller) external view override returns (uint256) {
+        return _pendingRedeemRequests[controller][requestId].amount;
+    }
+
+    /**
+     * @dev Gets the claimable redeem request amount.
+     */
+    function claimableRedeemRequest(uint256 requestId, address controller) external view override returns (uint256) {
+        return _pendingRedeemRequests[controller][requestId].claimable;
+    }
+
+    /**
+     * @dev Sets or revokes an operator for the given controller.
+     */
+    function setOperator(address operator, bool approved) external override returns (bool) {
+        _operators[msg.sender][operator] = approved;
+        emit OperatorSet(msg.sender, operator, approved);
+        return true;
+    }
+
+    /**
+     * @dev Checks if an operator is approved for a controller.
+     */
+    function isOperator(address controller, address operator) public view override returns (bool) {
+        return _operators[controller][operator];
+    }
+
+    /**
+     * @dev Internal function to generate a unique request ID.
+     */
+    function _generateRequestId(address controller, uint256 input) internal view returns (uint256) {
+        return uint256(keccak256(abi.encodePacked(block.timestamp, block.number, msg.sender, controller, input)));
```

Comment:
Using both `block.timestamp` and `block.number` is really a duplication of the same thing. We we ever expect one to change but not the other?

Note: on L2, block.number is more reliable.

Regardless, this function helps with uniqueness, but doesn't provide any real guarantee. While it is more expensive, maybe using a counter would be the better option.